c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,887 @@
+  //
+  // The native allocator's cacheInfo chooses to return the size of its largest
+  // unused block (which is the largest allocation the native allocator can
+  // service immediately and asynchronously without a cudaMalloc.
+  //
+  // Here, we use a different heuristic: figure out the max usable workspace
+  // size with a bit of educated trial and error. It's ok to be perf-inefficient
+  // because cacheInfo is a prelude to cudnnFind.
+  //
+  // The algo cache then stores the best-performing algo with workspace <=
+  // maxWorkspaceGuess. Later calls with the same param set hit in cache and try
+  // to allocate the same workspace. If, in one of those future calls, workspace
+  // allocation fails (ie because less ambient memory is available), the
+  // bindings rerun cudnnFind, including calling cacheInfo again beforehand to
+  // estimate a new (smaller) largest-available workspace. Over a few such
+  // calls, the cache should settle to the algo with a workspace size that's
+  // small enough to succeed every time (for that param set).
+  //
+  // So the strategy here is to return a rough, largeish guess and let the
+  // bindings retry to trim as needed over time.
+  //
+  // The only caveat is, even if a workspace is allocated without OOM errors now
+  // and in future calls, it's hard to be sure those later error-free
+  // cudaMallocAsyncs are fast and come straight from the pool (ie,
+  // cudaMallocAsync didn't need to reserve more memory from the system).
+  // Hopefully, after repeated workspace requests, the pool's reserved memory
+  // also stabilizes to a point where they all come straight from the pool.
+  std::lock_guard<std::mutex> lk(general_mutex);
+  assertValidDevice(device);
+  CUDAGuard g(device);
+  lazy_init_device(device);
+
+  size_t free_upper_bound;
+  size_t device_total;
+  C10_CUDA_CHECK(cudaMemGetInfo(&free_upper_bound, &device_total));
+  TORCH_INTERNAL_ASSERT(
+      free_upper_bound + pytorch_used_bytes[device] <= device_total);
+  size_t guess = std::min(
+      free_upper_bound,
+      pytorch_memory_limits[device] - pytorch_used_bytes[device]);
+  auto stream = c10::cuda::getCurrentCUDAStream();
+  void* dummy;
+
+  // Defensively checks for preexisting CUDA error state.
+  auto err = cudaGetLastError();
+  C10_CUDA_CHECK(err);
+
+  while (true) {
+    // Duplicates some logic from mallocAsync to work with the error state
+    // directly instead of repeatedly catching an exception thrown by
+    // mallocAsync.
+    if (pytorch_used_bytes[device] + guess > pytorch_memory_limits[device]) {
+      err = cudaErrorMemoryAllocation;
+    } else {
+      err = cudaMallocAsync(&dummy, guess, stream);
+    }
+
+    if (err == cudaSuccess) {
+      cudaFreeAsync(dummy, stream);
+      *maxWorkspaceGuess = guess;
+      return;
+    } else if (err == cudaErrorMemoryAllocation) {
+      cudaGetLastError(); // clear CUDA error
+      guess >>= 1; // quick and dirty: try half the size next iteration
```

Comment:
nit: there's no reason not to write this as `guess /= 2;` and let the compiler do strength reduction itself.